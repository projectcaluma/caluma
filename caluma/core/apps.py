from django.apps import AppConfig
from django.conf import settings
from django.utils.module_loading import import_string


class DefaultConfig(AppConfig):
    name = "caluma.core"

    def ready(self):
        from .mutation import Mutation
        from .types import Node
        from .serializers import ModelSerializer

        # to avoid recursive import error, load extension classes
        # only once the app is ready
        Mutation.permission_classes = [
            import_string(cls) for cls in settings.PERMISSION_CLASSES
        ]
        ModelSerializer.validation_classes = [
            import_string(cls) for cls in settings.VALIDATION_CLASSES
        ]
        Node.visibility_classes = [
            import_string(cls) for cls in settings.VISIBILITY_CLASSES
        ]
