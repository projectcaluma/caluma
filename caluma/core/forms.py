from graphene import List, String
from graphene_django import forms
from graphene_django.forms.converter import convert_form_field


class GlobalIDFormField(forms.GlobalIDFormField):
    """
    Global id form field which allows plain primary keys.

    Disable cleaning as input may be an actual relay global id
    or plain primary key.
    """

    def clean(self, value):
        return value


class GlobalIDMultipleChoiceField(forms.GlobalIDMultipleChoiceField):
    """
    Global id multiple choice field which allows plain primary keys.

    Disable cleaning as input may be an actual relay global id
    or plain primary key.
    """

    def valid_value(self, value):
        return True


class SlugMultipleChoiceField(GlobalIDMultipleChoiceField):
    """
    Slug multiple choice field which allows listing multiple slugs.

    Disable cleaning as we don't know which slugs may be valid down
    the line
    """

    def valid_value(self, value):  # pragma: no cover
        return True


@convert_form_field.register(SlugMultipleChoiceField)
def slug_to_list(field):
    return List(String, required=field.required)
