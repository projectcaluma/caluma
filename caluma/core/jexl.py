import pyjexl
from pyjexl.analysis import ValidatingAnalyzer
from pyjexl.exceptions import ParseError
from rest_framework import exceptions


class JexlValidator(object):
    def __init__(self, jexl):
        self.jexl = jexl

    def __call__(self, value):
        errors = list(self.jexl.validate(value))
        if errors:
            raise exceptions.ValidationError(errors)


class JEXL(pyjexl.JEXL):
    def validate(self, expression, ValidatingAnalyzerClass=ValidatingAnalyzer):
        try:
            for res in self.analyze(expression, ValidatingAnalyzerClass):
                yield res
        except ParseError as err:
            yield str(err)


class ExtractTransformSubjectAnalyzer(ValidatingAnalyzer):
    """
    Extract all subject values of given transforms.

    If no transforms are given all subjects of all transforms will be extracted.
    """

    def __init__(self, config, transforms=[]):
        self.transforms = transforms
        super().__init__(config)

    def visit_Transform(self, transform):
        if not self.transforms or transform.name in self.transforms:
            yield transform.subject.value
        yield from self.generic_visit(transform)
