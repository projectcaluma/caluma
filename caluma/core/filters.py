from functools import reduce

import django.forms
import graphene
from django import forms
from django.contrib.postgres.fields import JSONField
from django.contrib.postgres.fields.hstore import KeyTransform
from django.contrib.postgres.search import SearchVector
from django.db import models
from django.db.models.constants import LOOKUP_SEP
from django.db.models.functions import Cast
from django.utils import translation
from django_filters.constants import EMPTY_VALUES
from django_filters.fields import ChoiceField
from django_filters.rest_framework import (
    CharFilter,
    ChoiceFilter,
    Filter,
    FilterSet,
    MultipleChoiceFilter,
    OrderingFilter,
)
from graphene import Enum, InputObjectType, List
from graphene.types import generic
from graphene.types.utils import get_type
from graphene.utils.str_converters import to_camel_case
from graphene_django import filter
from graphene_django.converter import convert_choice_name
from graphene_django.filter.filterset import GrapheneFilterSetMixin
from graphene_django.forms.converter import convert_form_field
from graphene_django.registry import get_global_registry
from localized_fields.fields import LocalizedField

from .forms import GlobalIDFormField, GlobalIDMultipleChoiceField
from .relay import extract_global_id


class GlobalIDFilter(Filter):
    field_class = GlobalIDFormField

    def filter(self, qs, value):
        _id = None
        if value is not None:
            _id = extract_global_id(value)
        return super(GlobalIDFilter, self).filter(qs, _id)


class GlobalIDMultipleChoiceFilter(MultipleChoiceFilter):
    field_class = GlobalIDMultipleChoiceField

    def filter(self, qs, value):
        gids = [extract_global_id(v) for v in value]
        return super(GlobalIDMultipleChoiceFilter, self).filter(qs, gids)


class LocalizedFilter(Filter):
    def filter(self, qs, value):
        if value in EMPTY_VALUES:
            return qs

        lang = translation.get_language()
        filter_expr = "{0}__{1}__{2}".format(self.field_name, lang, self.lookup_expr)
        return qs.filter(**{filter_expr: value})


GrapheneFilterSetMixin.FILTER_DEFAULTS.update(
    {
        LocalizedField: {"filter_class": LocalizedFilter},
        models.AutoField: {"filter_class": GlobalIDFilter},
        models.OneToOneField: {"filter_class": GlobalIDFilter},
        models.ForeignKey: {"filter_class": GlobalIDFilter},
        models.ManyToManyField: {"filter_class": GlobalIDMultipleChoiceFilter},
        models.ManyToOneRel: {"filter_class": GlobalIDMultipleChoiceFilter},
        models.ManyToManyRel: {"filter_class": GlobalIDMultipleChoiceFilter},
    }
)


class SearchFilter(Filter):
    """
    Enable fulltext search on queryset.

    Define fields which need to be searched in.
    """

    def __init__(self, *args, fields, **kwargs):
        self.fields = fields
        super().__init__(*args, **kwargs)

    def _get_model_field(self, model, field):
        model_field = model._meta.get_field(field)
        return model_field, getattr(model_field, "related_model", None)

    def _build_search_expression(self, field_lookup):
        # TODO: is there no Django API which allows conversion of lookup to django field?
        model_field, _ = reduce(
            lambda model_tuple, field: self._get_model_field(model_tuple[1], field),
            field_lookup.split(LOOKUP_SEP),
            (None, self.model),
        )

        if isinstance(model_field, LocalizedField):
            lang = translation.get_language()
            return KeyTransform(lang, field_lookup)
        elif isinstance(model_field, JSONField):
            return Cast(field_lookup, models.TextField())

        return field_lookup

    def filter(self, qs, value):
        if value in EMPTY_VALUES:
            return qs

        qs = qs.annotate(
            search=SearchVector(
                *[self._build_search_expression(field) for field in self.fields]
            )
        )

        return qs.filter(search=value)


class OrderingField(ChoiceField):
    """
    Specific ordering field used as marker to convert to graphql type.

    See `convert_ordering_field_to_enum`
    """

    def validate(self, value):
        invalid = set(value or []) - {choice[0] for choice in self.choices}
        return not bool(invalid)

    def to_python(self, value):
        return value


class ListField(forms.Field):
    """List field as to allow actual lists in ordering vs csv string."""

    pass


class OrderingFilter(OrderingFilter):
    """Ordering filter adding default fields from models.BaseModel.

    Label is required and is used for enum naming in GraphQL schema.
    """

    base_field_class = ListField
    field_class = OrderingField

    def __init__(self, label, *args, fields=tuple(), **kwargs):
        fields = tuple(fields) + (
            "created_at",
            "modified_at",
            "created_by_user",
            "created_by_group",
        )

        super().__init__(
            *args,
            fields=fields,
            label=label,
            empty_label=None,
            null_label=None,
            **kwargs,
        )


class FilterSet(GrapheneFilterSetMixin, FilterSet):
    created_by_user = CharFilter()
    created_by_group = CharFilter()

    @classmethod
    def filter_for_lookup(cls, field, lookup_type):
        filter_class, params = super().filter_for_lookup(field, lookup_type)
        if issubclass(filter_class, ChoiceFilter):
            meta = field.model._meta
            # Postfixing newly created filter with Argument to avoid conflicts
            # with query nodes
            name = to_camel_case(f"{meta.object_name}_{field.name}_argument")
            params["label"] = name
        return filter_class, params


class MetaLookupMode(Enum):
    EXACT = "exact"
    STARTSWITH = "startswith"
    CONTAINS = "icontains"


class MetaValueFilterType(InputObjectType):
    key = graphene.String(required=True)
    value = generic.GenericScalar(required=True)
    lookup = MetaLookupMode()


class MetaValueFilterField(forms.MultiValueField):
    def __init__(self, label, **kwargs):
        super().__init__(fields=(forms.CharField(), forms.CharField()))

    def clean(self, data):
        # override parent clean() which would reject our data structure.
        # We don't validate, as the structure is already enforced by the
        # schema.
        return data


class MetaValueFilter(Filter):
    field_class = MetaValueFilterField

    def filter(self, qs, value):
        if value in EMPTY_VALUES:
            return qs
        meta_key = value["key"]
        meta_value = value["value"]
        lookup = value.get("lookup", self.lookup_expr)
        return qs.filter(**{f"{self.field_name}__{meta_key}__{lookup}": meta_value})


@convert_form_field.register(MetaValueFilterField)
def convert_meta_value_field(field):
    registry = get_global_registry()
    converted = registry.get_converted_field(field)
    if converted:
        return converted

    converted = MetaValueFilterType()
    registry.register_converted_field(field, converted)
    return converted


class MetaFilterSet(FilterSet):
    meta_has_key = CharFilter(lookup_expr="has_key", field_name="meta")
    meta_value = MetaValueFilter(field_name="meta")


class DjangoFilterConnectionField(filter.DjangoFilterConnectionField):
    """
    Django connection filter field with object type get_queryset support.

    Inspired by https://github.com/graphql-python/graphene-django/pull/528/files
    and might be removed once merged.
    """

    @property
    def filterset_class(self):
        return self._provided_filterset_class

    @classmethod
    def resolve_queryset(cls, connection, queryset, info, **args):
        return connection._meta.node.get_queryset(queryset, info)

    @classmethod
    def merge_querysets(cls, default_queryset, queryset):
        queryset = super().merge_querysets(default_queryset, queryset)
        # avoid query explosion of single relationships
        # may be removed once following issue is fixed:
        # https://github.com/graphql-python/graphene-django/issues/57
        return queryset.select_related()

    @classmethod
    def connection_resolver(
        cls,
        resolver,
        connection,
        default_manager,
        max_limit,
        enforce_first_or_last,
        filterset_class,
        filtering_args,
        root,
        info,
        **args,
    ):
        class QuerysetManager(object):
            def get_queryset(self):
                return cls.resolve_queryset(
                    connection, default_manager.get_queryset(), info, **args
                )

        return super().connection_resolver(
            resolver,
            connection,
            QuerysetManager(),
            max_limit,
            enforce_first_or_last,
            filterset_class,
            filtering_args,
            root,
            info,
            **args,
        )


class DjangoFilterSetConnectionField(DjangoFilterConnectionField):
    @property
    def model(self):
        return self.filterset_class._meta.model

    @property
    def type(self):
        return get_type(self._type)


@convert_form_field.register(OrderingField)
def convert_ordering_field_to_enum(field):
    """
    Add support to convert ordering choices to Graphql enum.

    Label is used as enum name.
    """
    registry = get_global_registry()
    converted = registry.get_converted_field(field)
    if converted:
        return converted

    def get_choices(choices):
        for value, help_text in choices:
            if value[0] != "-":
                name = convert_choice_name(value) + "_ASC"
            else:
                name = convert_choice_name(value[1:]) + "_DESC"
            description = help_text
            yield name, value, description

    name = to_camel_case(field.label)
    choices = list(get_choices(field.choices))
    named_choices = [(c[0], c[1]) for c in choices]
    named_choices_descriptions = {c[0]: c[2] for c in choices}

    class EnumWithDescriptionsType(object):
        @property
        def description(self):
            return named_choices_descriptions[self.name]

    enum = Enum(name, list(named_choices), type=EnumWithDescriptionsType)
    converted = List(enum, description=field.help_text, required=field.required)

    registry.register_converted_field(field, converted)
    return converted


@convert_form_field.register(ChoiceField)
def convert_choice_field_to_enum(field):
    """
    Add support to convert ordering choices to Graphql enum.

    Label is used as enum name.
    """

    registry = get_global_registry()
    converted = registry.get_converted_field(field)
    if converted:
        return converted

    def get_choices(choices):
        for value, help_text in choices:
            if value:
                name = convert_choice_name(value)
                description = help_text
                yield name, value, description

    name = to_camel_case(field.label)
    choices = list(get_choices(field.choices))
    named_choices = [(c[0], c[1]) for c in choices]
    named_choices_descriptions = {c[0]: c[2] for c in choices}

    class EnumWithDescriptionsType(object):
        @property
        def description(self):
            return named_choices_descriptions[self.name]

    enum = Enum(name, list(named_choices), type=EnumWithDescriptionsType)
    converted = enum(description=field.help_text, required=field.required)

    registry.register_converted_field(field, converted)
    return converted


def _generate_list_filter_class(inner_type):
    """
    Return a Filter class that will resolve into a List(`inner_type`) graphene type.

    This allows us to do things like use `__in` and `__overlap` filters that accept
    graphene lists instead of a comma delimited value string that's interpolated into
    a list by django_filters.BaseCSVFilter (which is used to define
    django_filters.BaseInFilter)
    """

    form_field = type(f"List{inner_type.__name__}FormField", (django.forms.Field,), {})
    filter_class = type(
        f"{inner_type.__name__}ListFilter",
        (Filter,),
        {
            "field_class": form_field,
            "__doc__": (
                f"{inner_type.__name__}ListFilter is a small extension of a raw "
                f"django_filters.Filter that allows us to express graphql "
                f"List({inner_type.__name__}) arguments using FilterSets. "
                f"Note that the given values are passed directly into queryset filters."
            ),
        },
    )
    convert_form_field.register(form_field)(
        lambda x: graphene.List(inner_type, required=x.required)
    )

    return filter_class


StringListFilter = _generate_list_filter_class(graphene.String)
