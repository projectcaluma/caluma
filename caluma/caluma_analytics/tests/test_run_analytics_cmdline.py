import json

import pytest
from django.core.management import call_command


@pytest.mark.parametrize(
    "enable_filter, expect_output", [(True, [0, 1]), (False, [0, 1, 2])]
)
@pytest.mark.parametrize("output_mode", [[], ["--sql"], ["--sqlonly"], ["--json"]])
def test_cmdline_output(
    db,
    settings,
    snapshot,
    form_and_document,
    case_factory,
    work_item_factory,
    task,
    form_question_factory,
    analytics_table,
    enable_filter,
    expect_output,
    output_mode,
    capsys,
):
    # two case/document combos that we will filter for
    form, document1, questions, answers_dict1 = form_and_document(False, False)
    case1 = case_factory(document=document1)
    form, document2, questions, answers_dict2 = form_and_document(False, False)
    case2 = case_factory(document=document2)

    # a third form/document/case combo that we don't filter for and should
    # therefore not be in the result set
    form, document3, questions, answers_dict3 = form_and_document(False, False)
    case3 = case_factory(document=document3)

    analytics_table.fields.create(
        alias="case_id",
        data_source="id",
    )
    analytics_table.fields.create(
        alias="blablub",
        data_source="document[top_form].top_question",
        filters=[
            answers_dict1["top_question"].value,
            answers_dict2["top_question"].value,
        ]
        if enable_filter
        else [],
    )

    args = [analytics_table.slug] + output_mode

    call_command("run_analytics", *args)

    out, err = capsys.readouterr()

    out = out.replace(str(case1.pk), "case1pk")
    out = out.replace(str(case2.pk), "case2pk")
    out = out.replace(str(case3.pk), "case3pk")

    if output_mode == ["--json"]:
        data = json.loads(out)
    else:
        data = out

    snapshot.assert_match({"data": data, "stderr": err})


def test_list_tables(db, analytics_table, capsys, snapshot):

    call_command("run_analytics")
    out, err = capsys.readouterr()

    assert analytics_table.slug in out
    snapshot.assert_match(out)
