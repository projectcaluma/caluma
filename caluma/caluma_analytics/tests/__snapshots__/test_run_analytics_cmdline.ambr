# name: test_cmdline_output[output_mode0-False-expect_output1]
  <class 'dict'> {
    'data': <class 'list'> [
      '------------------------------------   --------------  ',
      'case1pk   Shelly Watson   ',
      'case2pk   Daniel Stewart  ',
      'case3pk   Annette Mason   ',
      'case_id                                blablub         ',
    ],
    'stderr': '',
  }
---
# name: test_cmdline_output[output_mode0-True-expect_output0]
  <class 'dict'> {
    'data': <class 'list'> [
      '------------------------------------   --------------  ',
      'case1pk   Shelly Watson   ',
      'case2pk   Daniel Stewart  ',
      'case_id                                blablub         ',
    ],
    'stderr': '',
  }
---
# name: test_cmdline_output[output_mode1-False-expect_output1]
  <class 'dict'> {
    'data': <class 'list'> [
      '------------------------------------   --------------  ',
      'case1pk   Shelly Watson   ',
      'case2pk   Daniel Stewart  ',
      'case3pk   Annette Mason   ',
      'case_id                                blablub         ',
    ],
    'stderr': '
      -- SQL: 
      WITH
          case_ac50e AS (SELECT "caluma_workflow_case"."created_at",  
          "caluma_workflow_case"."modified_at",  
          "caluma_workflow_case"."created_by_user",  
          "caluma_workflow_case"."created_by_group",  
          "caluma_workflow_case"."modified_by_user",  
          "caluma_workflow_case"."modified_by_group",  
          "caluma_workflow_case"."id",  
          "caluma_workflow_case"."family_id",  
          "caluma_workflow_case"."closed_at",  
          "caluma_workflow_case"."closed_by_user",  
          "caluma_workflow_case"."closed_by_group",  
          "caluma_workflow_case"."workflow_id",  
          "caluma_workflow_case"."status",  
          "caluma_workflow_case"."meta",  
          "caluma_workflow_case"."document_id"
          FROM
                 "caluma_workflow_case" INNER JOIN "caluma_workflow_workflow" ON ("caluma_workflow_case"."workflow_id" = "caluma_workflow_workflow"."slug") -- qs ref
          ),
          document_2a07e AS (SELECT "caluma_form_document"."created_at",  
          "caluma_form_document"."modified_at",  
          "caluma_form_document"."created_by_user",  
          "caluma_form_document"."created_by_group",  
          "caluma_form_document"."modified_by_user",  
          "caluma_form_document"."modified_by_group",  
          "caluma_form_document"."id",  
          "caluma_form_document"."family_id",  
          "caluma_form_document"."form_id",  
          "caluma_form_document"."source_id",  
          "caluma_form_document"."meta"
          FROM
                 "caluma_form_document" INNER JOIN "caluma_form_form" ON ("caluma_form_document"."form_id" = "caluma_form_form"."slug") -- qs ref
          ),
          answer_b4215 AS (SELECT "caluma_form_answer"."created_at",  
          "caluma_form_answer"."modified_at",  
          "caluma_form_answer"."created_by_user",  
          "caluma_form_answer"."created_by_group",  
          "caluma_form_answer"."modified_by_user",  
          "caluma_form_answer"."modified_by_group",  
          "caluma_form_answer"."id",  
          "caluma_form_answer"."question_id",  
          "caluma_form_answer"."value",  
          "caluma_form_answer"."meta",  
          "caluma_form_answer"."document_id",  
          "caluma_form_answer"."date",  
          "caluma_form_answer"."file_id"
          FROM
                 "caluma_form_answer" INNER JOIN "caluma_form_question" ON ("caluma_form_answer"."question_id" = "caluma_form_question"."slug") -- qs ref
          )
      SELECT * FROM (SELECT 
             case_ac50e.id AS "analytics_result_case_id",
             analytics_result_blablub
      FROM case_ac50e AS case_ac50e 
      LEFT JOIN (
          SELECT DISTINCT ON (id)
                 id,
                 analytics_result_blablub
          FROM document_2a07e AS document_2a07e 
          LEFT JOIN (
                  SELECT DISTINCT ON (document_id)
                         (answer_b4215.value #>>'{}') AS "analytics_result_blablub",
                         document_id
                  FROM answer_b4215 AS answer_b4215 
                  WHERE "question_id" = 'top_question' 
                  ORDER BY document_id
      
          ) AS answer_b4215_caea5 ON (document_2a07e.id = "answer_b4215_caea5".document_id)
          WHERE form_id = 'top_form'
          ORDER BY id
      
      ) AS document_2a07e_27f15 ON (case_ac50e.document_id = "document_2a07e_27f15".id)
      
      ) AS analytics_76b97
      -- PARAMS: 
  
    ',
  }
---
# name: test_cmdline_output[output_mode1-True-expect_output0]
  <class 'dict'> {
    'data': <class 'list'> [
      '------------------------------------   --------------  ',
      'case1pk   Shelly Watson   ',
      'case2pk   Daniel Stewart  ',
      'case_id                                blablub         ',
    ],
    'stderr': '
      -- SQL: 
      WITH
          case_ac50e AS (SELECT "caluma_workflow_case"."created_at",  
          "caluma_workflow_case"."modified_at",  
          "caluma_workflow_case"."created_by_user",  
          "caluma_workflow_case"."created_by_group",  
          "caluma_workflow_case"."modified_by_user",  
          "caluma_workflow_case"."modified_by_group",  
          "caluma_workflow_case"."id",  
          "caluma_workflow_case"."family_id",  
          "caluma_workflow_case"."closed_at",  
          "caluma_workflow_case"."closed_by_user",  
          "caluma_workflow_case"."closed_by_group",  
          "caluma_workflow_case"."workflow_id",  
          "caluma_workflow_case"."status",  
          "caluma_workflow_case"."meta",  
          "caluma_workflow_case"."document_id"
          FROM
                 "caluma_workflow_case" INNER JOIN "caluma_workflow_workflow" ON ("caluma_workflow_case"."workflow_id" = "caluma_workflow_workflow"."slug") -- qs ref
          ),
          document_2a07e AS (SELECT "caluma_form_document"."created_at",  
          "caluma_form_document"."modified_at",  
          "caluma_form_document"."created_by_user",  
          "caluma_form_document"."created_by_group",  
          "caluma_form_document"."modified_by_user",  
          "caluma_form_document"."modified_by_group",  
          "caluma_form_document"."id",  
          "caluma_form_document"."family_id",  
          "caluma_form_document"."form_id",  
          "caluma_form_document"."source_id",  
          "caluma_form_document"."meta"
          FROM
                 "caluma_form_document" INNER JOIN "caluma_form_form" ON ("caluma_form_document"."form_id" = "caluma_form_form"."slug") -- qs ref
          ),
          answer_b4215 AS (SELECT "caluma_form_answer"."created_at",  
          "caluma_form_answer"."modified_at",  
          "caluma_form_answer"."created_by_user",  
          "caluma_form_answer"."created_by_group",  
          "caluma_form_answer"."modified_by_user",  
          "caluma_form_answer"."modified_by_group",  
          "caluma_form_answer"."id",  
          "caluma_form_answer"."question_id",  
          "caluma_form_answer"."value",  
          "caluma_form_answer"."meta",  
          "caluma_form_answer"."document_id",  
          "caluma_form_answer"."date",  
          "caluma_form_answer"."file_id"
          FROM
                 "caluma_form_answer" INNER JOIN "caluma_form_question" ON ("caluma_form_answer"."question_id" = "caluma_form_question"."slug") -- qs ref
          )
      SELECT * FROM (SELECT 
             case_ac50e.id AS "analytics_result_case_id",
             analytics_result_blablub
      FROM case_ac50e AS case_ac50e 
      LEFT JOIN (
          SELECT DISTINCT ON (id)
                 id,
                 analytics_result_blablub
          FROM document_2a07e AS document_2a07e 
          LEFT JOIN (
                  SELECT DISTINCT ON (document_id)
                         (answer_b4215.value #>>'{}') AS "analytics_result_blablub",
                         document_id
                  FROM answer_b4215 AS answer_b4215 
                  WHERE "question_id" = 'top_question' 
                  ORDER BY document_id
      
          ) AS answer_b4215_caea5 ON (document_2a07e.id = "answer_b4215_caea5".document_id)
          WHERE form_id = 'top_form'
          ORDER BY id
      
      ) AS document_2a07e_27f15 ON (case_ac50e.document_id = "document_2a07e_27f15".id)
      
      ) AS analytics_76b97
      WHERE analytics_result_blablub IN (%(flt_analytics_result_blablub_8201d)s, %(flt_analytics_result_blablub_8e5e6)s)
      -- PARAMS: 
      --     flt_analytics_result_blablub_8201d: Shelly Watson
      --     flt_analytics_result_blablub_8e5e6: Daniel Stewart
  
    ',
  }
---
# name: test_cmdline_output[output_mode2-False-expect_output1]
  <class 'dict'> {
    'data': <class 'list'> [
    ],
    'stderr': '
      -- SQL: 
      WITH
          case_ac50e AS (SELECT "caluma_workflow_case"."created_at",  
          "caluma_workflow_case"."modified_at",  
          "caluma_workflow_case"."created_by_user",  
          "caluma_workflow_case"."created_by_group",  
          "caluma_workflow_case"."modified_by_user",  
          "caluma_workflow_case"."modified_by_group",  
          "caluma_workflow_case"."id",  
          "caluma_workflow_case"."family_id",  
          "caluma_workflow_case"."closed_at",  
          "caluma_workflow_case"."closed_by_user",  
          "caluma_workflow_case"."closed_by_group",  
          "caluma_workflow_case"."workflow_id",  
          "caluma_workflow_case"."status",  
          "caluma_workflow_case"."meta",  
          "caluma_workflow_case"."document_id"
          FROM
                 "caluma_workflow_case" INNER JOIN "caluma_workflow_workflow" ON ("caluma_workflow_case"."workflow_id" = "caluma_workflow_workflow"."slug") -- qs ref
          ),
          document_2a07e AS (SELECT "caluma_form_document"."created_at",  
          "caluma_form_document"."modified_at",  
          "caluma_form_document"."created_by_user",  
          "caluma_form_document"."created_by_group",  
          "caluma_form_document"."modified_by_user",  
          "caluma_form_document"."modified_by_group",  
          "caluma_form_document"."id",  
          "caluma_form_document"."family_id",  
          "caluma_form_document"."form_id",  
          "caluma_form_document"."source_id",  
          "caluma_form_document"."meta"
          FROM
                 "caluma_form_document" INNER JOIN "caluma_form_form" ON ("caluma_form_document"."form_id" = "caluma_form_form"."slug") -- qs ref
          ),
          answer_b4215 AS (SELECT "caluma_form_answer"."created_at",  
          "caluma_form_answer"."modified_at",  
          "caluma_form_answer"."created_by_user",  
          "caluma_form_answer"."created_by_group",  
          "caluma_form_answer"."modified_by_user",  
          "caluma_form_answer"."modified_by_group",  
          "caluma_form_answer"."id",  
          "caluma_form_answer"."question_id",  
          "caluma_form_answer"."value",  
          "caluma_form_answer"."meta",  
          "caluma_form_answer"."document_id",  
          "caluma_form_answer"."date",  
          "caluma_form_answer"."file_id"
          FROM
                 "caluma_form_answer" INNER JOIN "caluma_form_question" ON ("caluma_form_answer"."question_id" = "caluma_form_question"."slug") -- qs ref
          )
      SELECT * FROM (SELECT 
             case_ac50e.id AS "analytics_result_case_id",
             analytics_result_blablub
      FROM case_ac50e AS case_ac50e 
      LEFT JOIN (
          SELECT DISTINCT ON (id)
                 id,
                 analytics_result_blablub
          FROM document_2a07e AS document_2a07e 
          LEFT JOIN (
                  SELECT DISTINCT ON (document_id)
                         (answer_b4215.value #>>'{}') AS "analytics_result_blablub",
                         document_id
                  FROM answer_b4215 AS answer_b4215 
                  WHERE "question_id" = 'top_question' 
                  ORDER BY document_id
      
          ) AS answer_b4215_caea5 ON (document_2a07e.id = "answer_b4215_caea5".document_id)
          WHERE form_id = 'top_form'
          ORDER BY id
      
      ) AS document_2a07e_27f15 ON (case_ac50e.document_id = "document_2a07e_27f15".id)
      
      ) AS analytics_76b97
      -- PARAMS: 
  
    ',
  }
---
# name: test_cmdline_output[output_mode2-True-expect_output0]
  <class 'dict'> {
    'data': <class 'list'> [
    ],
    'stderr': '
      -- SQL: 
      WITH
          case_ac50e AS (SELECT "caluma_workflow_case"."created_at",  
          "caluma_workflow_case"."modified_at",  
          "caluma_workflow_case"."created_by_user",  
          "caluma_workflow_case"."created_by_group",  
          "caluma_workflow_case"."modified_by_user",  
          "caluma_workflow_case"."modified_by_group",  
          "caluma_workflow_case"."id",  
          "caluma_workflow_case"."family_id",  
          "caluma_workflow_case"."closed_at",  
          "caluma_workflow_case"."closed_by_user",  
          "caluma_workflow_case"."closed_by_group",  
          "caluma_workflow_case"."workflow_id",  
          "caluma_workflow_case"."status",  
          "caluma_workflow_case"."meta",  
          "caluma_workflow_case"."document_id"
          FROM
                 "caluma_workflow_case" INNER JOIN "caluma_workflow_workflow" ON ("caluma_workflow_case"."workflow_id" = "caluma_workflow_workflow"."slug") -- qs ref
          ),
          document_2a07e AS (SELECT "caluma_form_document"."created_at",  
          "caluma_form_document"."modified_at",  
          "caluma_form_document"."created_by_user",  
          "caluma_form_document"."created_by_group",  
          "caluma_form_document"."modified_by_user",  
          "caluma_form_document"."modified_by_group",  
          "caluma_form_document"."id",  
          "caluma_form_document"."family_id",  
          "caluma_form_document"."form_id",  
          "caluma_form_document"."source_id",  
          "caluma_form_document"."meta"
          FROM
                 "caluma_form_document" INNER JOIN "caluma_form_form" ON ("caluma_form_document"."form_id" = "caluma_form_form"."slug") -- qs ref
          ),
          answer_b4215 AS (SELECT "caluma_form_answer"."created_at",  
          "caluma_form_answer"."modified_at",  
          "caluma_form_answer"."created_by_user",  
          "caluma_form_answer"."created_by_group",  
          "caluma_form_answer"."modified_by_user",  
          "caluma_form_answer"."modified_by_group",  
          "caluma_form_answer"."id",  
          "caluma_form_answer"."question_id",  
          "caluma_form_answer"."value",  
          "caluma_form_answer"."meta",  
          "caluma_form_answer"."document_id",  
          "caluma_form_answer"."date",  
          "caluma_form_answer"."file_id"
          FROM
                 "caluma_form_answer" INNER JOIN "caluma_form_question" ON ("caluma_form_answer"."question_id" = "caluma_form_question"."slug") -- qs ref
          )
      SELECT * FROM (SELECT 
             case_ac50e.id AS "analytics_result_case_id",
             analytics_result_blablub
      FROM case_ac50e AS case_ac50e 
      LEFT JOIN (
          SELECT DISTINCT ON (id)
                 id,
                 analytics_result_blablub
          FROM document_2a07e AS document_2a07e 
          LEFT JOIN (
                  SELECT DISTINCT ON (document_id)
                         (answer_b4215.value #>>'{}') AS "analytics_result_blablub",
                         document_id
                  FROM answer_b4215 AS answer_b4215 
                  WHERE "question_id" = 'top_question' 
                  ORDER BY document_id
      
          ) AS answer_b4215_caea5 ON (document_2a07e.id = "answer_b4215_caea5".document_id)
          WHERE form_id = 'top_form'
          ORDER BY id
      
      ) AS document_2a07e_27f15 ON (case_ac50e.document_id = "document_2a07e_27f15".id)
      
      ) AS analytics_76b97
      WHERE analytics_result_blablub IN (%(flt_analytics_result_blablub_8201d)s, %(flt_analytics_result_blablub_8e5e6)s)
      -- PARAMS: 
      --     flt_analytics_result_blablub_8201d: Shelly Watson
      --     flt_analytics_result_blablub_8e5e6: Daniel Stewart
  
    ',
  }
---
# name: test_cmdline_output[output_mode3-False-expect_output1]
  <class 'dict'> {
    'data': <class 'list'> [
      <class 'dict'> {
        'blablub': 'Annette Mason',
        'case_id': 'case3pk',
      },
      <class 'dict'> {
        'blablub': 'Daniel Stewart',
        'case_id': 'case2pk',
      },
      <class 'dict'> {
        'blablub': 'Shelly Watson',
        'case_id': 'case1pk',
      },
    ],
    'stderr': '',
  }
---
# name: test_cmdline_output[output_mode3-True-expect_output0]
  <class 'dict'> {
    'data': <class 'list'> [
      <class 'dict'> {
        'blablub': 'Daniel Stewart',
        'case_id': 'case2pk',
      },
      <class 'dict'> {
        'blablub': 'Shelly Watson',
        'case_id': 'case1pk',
      },
    ],
    'stderr': '',
  }
---
# name: test_list_tables
  '
  No analytics table specified. The following tables are available:
    ID:                  NAME:
    environmental-ten    Bonnie Moreno
  
  '
---
